### DISCUSSION TITLE
Are the .ps1 and psm1 modules needed?

### DISCUSSION THREAD

**Discussion by marticliment** [gerardog/gsudo#166] [Q&A] [answered]
Hi,

I'm trying to integrate Gsudo on an app, and I wonder if I really need the .ps1, psm1 and the `gsudo` file if I only call `gsudo.exe` with a system call.

Thanks!

**Comment by gerardog** [comment, marked as answer]
If you deploy just the `.exe`, then your `cmd` elevations will work fine, also `gsudo !!` (when called from cmd) will work. If you call `gsudo` from your app, it will work.

- `gsudo` (the one without extension) is used to properly elevate bash commands from inside wsl / git-bash / cygwin / etc ... 
- `Invoke-gsudo.ps1` provides the Invoke-gsudo PowerShell cmdlet (Nicer powershell syntax).
- `gsudoModule.psm1` and `gsudoModule.psd1` provides the PowerShell module that, if Imported in the profile, makes `gsudo !!` work from PowerShell and also Invoke-gsudo AutoComplete. 

Those files are not really required if you don't do those things, but they improve the user experience... If lacking, the user would either think the documentation is wrong or the product is flawed.

I would say that, if you planned that your app would put gsudo in the user `$path`, allowing the user to call gsudo from any command line, then you'd rather put the full bundle, or trigger a `gsudo` install with one of the provided installations methods. For example, running: `PowerShell -Command "Set-ExecutionPolicy RemoteSigned -scope Process; [Net.ServicePointManager]::SecurityProtocol = 'Tls12'; iwr -useb https://raw.githubusercontent.com/gerardog/gsudo/master/installgsudo.ps1 | iex"`

If you only want to invoke `gsudo {cmd command}`, then, you could use just deliver the `gsudo.exe`... But if that is the case, have you considered CreateProcess with `runas` verb? it elevates in a new window, but does not requires any third party tool.

Example C# 
``` c#
ProcessStartInfo startInfo = default(ProcessStartInfo);

startInfo = new ProcessStartInfo(SetupProgramPath)
{
    UseShellExecute = true,
    Verb = "runas",
    WindowStyle = ProcessWindowStyle.Normal,
    FileName = "msiexec",
    Arguments = "/i \"" + SetupProgramPath + "\"",
    CreateNoWindow = false
};

Process.Start(startInfo); 
```

Example PowerShell 
``` powershell
Start-Process -FilePath "powershell.exe" -Verb RunAs
```

**Reply by marticliment** [reply]
> But if that is the case, have you considered CreateProcess with runas verb? it elevates in a new window, but does not requires any third party tool

Yes, but it does not work as expected🤷

**Reply by gerardog** [reply]
What are you expecting and what are the results?

**Reply by marticliment** [reply]
I was planning to be able to get the output of the command, as well as wait for it (but not the spawned subprocesses) to finish from Python, and Start-process wouldn't allow me. (due to security resaons i guess?)

**Reply by marticliment** [reply]
If you are curious, it is just here https://github.com/martinet101/WingetUI/. I needed a way to force admin installs, and powershell wouldn't work

**Reply by gerardog** [reply]
You may elevate (CreateProcess with verb runas) something like "cmd /c {your command} > %temp%\output5342543.txt", and also wait for the process to end 

![image](https://user-images.githubusercontent.com/3901474/188437590-37ddeb0e-eff7-4bcb-b60d-0e72a9097fc7.png)

In python, maybe [https://stackoverflow.com/questions/19672352/how-to-run-script-with-elevated-privilege-on-windows/19719292#19719292](this) (see the `if wait`) or 'https://pypi.org/project/elevate/', or 'gsudo', whatever fits.

**Reply by marticliment** [reply]
Yes, that could be an option, but with gsudo I get the live output and I can communicate with the command line
